hub/services/init_target_cluster_sub_step.go

```
@@ -131,32 +131,99 @@ func upgradeDataDir(path string) string {
-	master.DataDir = upgradeDataDir(master.DataDir)
-	datadirDeclare := fmt.Sprintf("QD_PRIMARY_ARRAY=%s~%d~%s~%d~%d~0",
-		master.Hostname, master.Port, master.DataDir, master.DbID, master.ContentID)
-	gpinitsystemConfig = append(gpinitsystemConfig, datadirDeclare)
-
-	// declare segment data directories
-	segmentDeclarations := []string{}
+// sanitize sorts and deduplicates a slice of port numbers.
+func sanitize(ports []uint32) []uint32 {
+	sort.Slice(ports, func(i, j int) bool { return ports[i] < ports[j] })
+
+	dedupe := ports[:0] // point at the same backing array
+
+	var last uint32
+	for i, port := range ports {
+		if i == 0 || port != last {
+			dedupe = append(dedupe, port)
+		}
+		last = port
+	}
+
+	return dedupe
+}
+
+func WriteSegmentArray(config []string, source *utils.Cluster, ports []uint32) ([]string, int, error) {
+	// Partition segments by host in order to correctly assign ports.
+	segmentsByHost := make(map[string][]cluster.SegConfig)
 	for _, content := range source.ContentIDs {
-		if content != -1 {
-			segment := source.Segments[content]
-			// FIXME: Arbitrary assumption.	 Do something smarter later
-			segment.Port += 4000
-			segment.DataDir = upgradeDataDir(segment.DataDir)
-			declaration := fmt.Sprintf("\t%s~%d~%s~%d~%d~0",
-				segment.Hostname, segment.Port, segment.DataDir, segment.DbID, segment.ContentID)
-			segmentDeclarations = append(segmentDeclarations, declaration)
+		segment := source.Segments[content]
+		segmentsByHost[segment.Hostname] = append(segmentsByHost[segment.Hostname], segment)
+	}
+
+	if len(ports) == 0 {
+		// Create a default port range, starting with the pg_upgrade default of
+		// 50432. Reserve enough ports to handle the host with the most
+		// segments.
+		var maxSegs int
+		for _, segments := range segmentsByHost {
+			if len(segments) > maxSegs {
+				maxSegs = len(segments)
+			}
 		}
+
+		for i := 0; i < maxSegs; i++ {
+			ports = append(ports, uint32(50432+i))
+		}
+	}
+
+	ports = sanitize(ports)
```

I'm thinking `sanitize` should be done in the CLI as part of the input checking instead.  If a user specifies the same port multiple times, it's almost certainly an unintended user error, and we should fail fast so they can correct it.  I see no reason to permit a user to specify the same port multiple times. Presumably the database initialization code will fail if it tries to use the same port twice, and that is our real check here.

Also, `sanitize` makes sense as a concept, but all we are doing in that function here is deduplication.  Since we seem to be deferring all real checking of the ports to `gpinitsystem`, perhaps a CLI function called `checkPorts` or similar would make more sense.

---

> If a user specifies the same port multiple times, it's almost certainly an unintended user error, and we should fail fast so they can correct it.

`1-5,5-10` is unambiguous but still needs to be deduplicated for safety. There's not much reason to make the user jump through hoops here IMO.

> Also, sanitize makes sense as a concept, but all we are doing in that function here is deduplication.

And sorting, which is critical for the implementation.

---

`1-5,5-10` is fine, but what about `6000,6001,6001,6003`(that is, a typo)?  

It is really a UI question, and my (relatively uninformed UI opinion) is that we should force the user to give us an unambiguous list. 

This need not delay the PR, but if we cannot receive some UI input from others, we should at least add an issue to rethink this later.

---

> 1-5,5-10 is fine, but what about 6000,6001,6001,6003(that is, a typo)?

My arguments are
1. There's no difference from the internal perspective.
2. Users are likely to specify the latter as `6000-6003` anyway.
3. The penalty for a typo here is low-impact. If they don't have enough ports due to the duplication, we tell them within seconds. If they do have enough ports, we're still using the range they provided; it's just slightly smaller than they thought. And this temporary range is only used during upgrade.

---

(@kodonnell26 , in case David's ping didn't go through on the edit.)

---

Though after `execute`, they can have both clusters up for a while....

---

> Though after execute, they can have both clusters up for a while....

Sure. What negative consequences do you see happening in this situation?

---

Is there a reason we're specifying a port range and not just a PORT_BASE like we do in `gpinitsystem`?

---

Ok, after discussion with Brendan from Support, repeated ports are fine as long as there are enough of them.

@bharath1028 was wondering, however, why we chose the current explicit scheme(full range for segments) instead of specifying a single `PORT_BASE` that [gpinitsystem](https://gpdb.docs.pivotal.io/6-1/utility_guide/ref/gpinitsystem.html) allows.

---

@bharath1028 @dkrieger-pvtl Well, this is embarrassing. The original intent was to make sure that we didn't have to reserve a port range from `PORT_BASE`-`PORT_BASE + <number of segments>`. But it turns out that the current `PORT_BASE` scheme already does that. Apparently I have been holding on to that misconception for a long time.

The current scheme, then, only has the advantage that the temporary ports are not required to be contiguous. I still _like_ that, but I admit it's a much weaker justification.

---

To me, we don't need to finalize on this decision before this PR goes in...the additional flexibility of a non-contiguous port range _can_ be removed later as a product decision.  But I think we need more customer input before finalizing this.  So, provided we add an issue to revisit this decision before release, the current scheme is fine with me.

---

+1 @dkrieger-pvtl. Thanks for the heads up @jchampio!

---

I also understand and agree that `sanitize` should go into the hub code for security considerations.